### ISSUE TITLE
tcpreplay4.3.1, use "tcprewrite --fixlen" to fix packet length, error in realloc()

### ISSUE THREAD

**Issue by lumin659** [appneta/tcpreplay#541]
[root@localhost test]# tcprewrite --fixlen=pad -C -c 187302.cache -i 187302.pcap -o 187302-new.pcap
Warning: 187302.pcap was captured using a snaplen of 2048 bytes.  This may mean you have truncated packets.
*** Error in `tcprewrite': realloc(): invalid next size: 0x000000000093c0b0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x7f5d4)[0x7f7d008cf5d4]
/lib64/libc.so.6(+0x83bd8)[0x7f7d008d3bd8]
/lib64/libc.so.6(realloc+0x1d2)[0x7f7d008d5832]
tcprewrite[0x410d7b]
tcprewrite[0x405f0d]
tcprewrite[0x40429c]
tcprewrite[0x403c3c]
tcprewrite[0x402f3e]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f7d008723d5]
tcprewrite[0x403087]
======= Memory map: ========
00400000-00436000 r-xp 00000000 fd:00 9267872                            /usr/local/bin/tcprewrite
00636000-00637000 r--p 00036000 fd:00 9267872                            /usr/local/bin/tcprewrite
00637000-0063a000 rw-p 00037000 fd:00 9267872                            /usr/local/bin/tcprewrite
00934000-0097d000 rw-p 00000000 00:00 0                                  [heap]
7f7cfc000000-7f7cfc021000 rw-p 00000000 00:00 0 
7f7cfc021000-7f7d00000000 ---p 00000000 00:00 0 
7f7d0041e000-7f7d00433000 r-xp 00000000 fd:00 23684                      /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7f7d00433000-7f7d00632000 ---p 00015000 fd:00 23684                      /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7f7d00632000-7f7d00633000 r--p 00014000 fd:00 23684                      /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7f7d00633000-7f7d00634000 rw-p 00015000 fd:00 23684                      /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7f7d00634000-7f7d0064b000 r-xp 00000000 fd:00 907114                     /usr/lib64/libpthread-2.17.so
7f7d0064b000-7f7d0084a000 ---p 00017000 fd:00 907114                     /usr/lib64/libpthread-2.17.so
7f7d0084a000-7f7d0084b000 r--p 00016000 fd:00 907114                     /usr/lib64/libpthread-2.17.so
7f7d0084b000-7f7d0084c000 rw-p 00017000 fd:00 907114                     /usr/lib64/libpthread-2.17.so
7f7d0084c000-7f7d00850000 rw-p 00000000 00:00 0 
7f7d00850000-7f7d00a12000 r-xp 00000000 fd:00 33133                      /usr/lib64/libc-2.17.so
7f7d00a12000-7f7d00c12000 ---p 001c2000 fd:00 33133                      /usr/lib64/libc-2.17.so
7f7d00c12000-7f7d00c16000 r--p 001c2000 fd:00 33133                      /usr/lib64/libc-2.17.so
7f7d00c16000-7f7d00c18000 rw-p 001c6000 fd:00 33133                      /usr/lib64/libc-2.17.so
7f7d00c18000-7f7d00c1d000 rw-p 00000000 00:00 0 
7f7d00c1d000-7f7d00c33000 r-xp 00000000 fd:00 33147                      /usr/lib64/libnsl-2.17.so
7f7d00c33000-7f7d00e33000 ---p 00016000 fd:00 33147                      /usr/lib64/libnsl-2.17.so
7f7d00e33000-7f7d00e34000 r--p 00016000 fd:00 33147                      /usr/lib64/libnsl-2.17.so
7f7d00e34000-7f7d00e35000 rw-p 00017000 fd:00 33147                      /usr/lib64/libnsl-2.17.so
7f7d00e35000-7f7d00e37000 rw-p 00000000 00:00 0 
7f7d00e37000-7f7d00e3e000 r-xp 00000000 fd:00 907118                     /usr/lib64/librt-2.17.so
7f7d00e3e000-7f7d0103d000 ---p 00007000 fd:00 907118                     /usr/lib64/librt-2.17.so
7f7d0103d000-7f7d0103e000 r--p 00006000 fd:00 907118                     /usr/lib64/librt-2.17.so
7f7d0103e000-7f7d0103f000 rw-p 00007000 fd:00 907118                     /usr/lib64/librt-2.17.so
7f7d0103f000-7f7d0104c000 r-xp 00000000 fd:00 360893                     /usr/lib64/libdnet.so.1.0.1
7f7d0104c000-7f7d0124b000 ---p 0000d000 fd:00 360893                     /usr/lib64/libdnet.so.1.0.1
7f7d0124b000-7f7d0124d000 r--p 0000c000 fd:00 360893                     /usr/lib64/libdnet.so.1.0.1
7f7d0124d000-7f7d0124e000 rw-p 0000e000 fd:00 360893                     /usr/lib64/libdnet.so.1.0.1
7f7d0124e000-7f7d01250000 rw-p 00000000 00:00 0 
7f7d01250000-7f7d0128e000 r-xp 00000000 fd:00 784683                     /usr/lib64/libpcap.so.1.5.3
7f7d0128e000-7f7d0148d000 ---p 0003e000 fd:00 784683                     /usr/lib64/libpcap.so.1.5.3
7f7d0148d000-7f7d0148f000 r--p 0003d000 fd:00 784683                     /usr/lib64/libpcap.so.1.5.3
7f7d0148f000-7f7d01490000 rw-p 0003f000 fd:00 784683                     /usr/lib64/libpcap.so.1.5.3
7f7d01490000-7f7d01491000 rw-p 00000000 00:00 0 
7f7d01491000-7f7d014b3000 r-xp 00000000 fd:00 907097                     /usr/lib64/ld-2.17.so
7f7d016a1000-7f7d016a6000 rw-p 00000000 00:00 0 
7f7d016ae000-7f7d016b2000 rw-p 00000000 00:00 0 
7f7d016b2000-7f7d016b3000 r--p 00021000 fd:00 907097                     /usr/lib64/ld-2.17.so
7f7d016b3000-7f7d016b4000 rw-p 00022000 fd:00 907097                     /usr/lib64/ld-2.17.so
7f7d016b4000-7f7d016b5000 rw-p 00000000 00:00 0 
7ffdd39ba000-7ffdd39db000 rw-p 00000000 00:00 0                          [stack]
7ffdd39f1000-7ffdd39f4000 r--p 00000000 00:00 0                          [vvar]
7ffdd39f4000-7ffdd39f6000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
Aborted (core dumped)

**Comment by fklassen** [comment]
Please provide packet capture and cache file.

**Comment by fcoiffie** [comment]
I can reproduce this problem with this command and the attached pcap:
```
tcprewrite -i capture.pcap -F pad  -o out.pcap
```
[capture.pcap.gz](https://github.com/appneta/tcpreplay/files/3222649/capture.pcap.gz)

**Comment by fcoiffie** [comment]
Better but not enough to solve all the problems (cf Valgrind output):
```diff
diff --git a/src/tcpedit/tcpedit.c b/src/tcpedit/tcpedit.c
index 69659dfa..89ea878c 100644
--- a/src/tcpedit/tcpedit.c
+++ b/src/tcpedit/tcpedit.c
@@ -279,8 +279,17 @@ tcpedit_packet(tcpedit_t *tcpedit, struct pcap_pkthdr **pkthdr,
 
     /* (Un)truncate or MTU truncate packet? */
     if (tcpedit->fixlen || tcpedit->mtu_truncate) {
+        size_t offset = (u_char*)ip_hdr - *pktdata;
+        size_t offset6 = (u_char*)ip6_hdr - *pktdata;
         if ((retval = untrunc_packet(tcpedit, *pkthdr, pktdata, ip_hdr, ip6_hdr)) < 0)
             return TCPEDIT_ERROR;
+        if (ip_hdr) {
+            ip_hdr = *pktdata + offset;
+        }
+        if (ip6_hdr) {
+            ip6_hdr = *pktdata + offset6;
+        }
+        packet = *pktdata;
         needtorecalc += retval;
     }
```

**Comment by fklassen** [comment]
Fixed in #594
